src/server/coprocessor.rs

```diff
@@ -106,15 +108,46 @@ impl EndPointHost {
             pool: ThreadPool::new(DEFAULT_POOL_SIZE),
         }
     }
+}
 
-    pub fn on_request(&self, req: Request, on_resp: OnResponse) {
-        let end_point = self.snap_endpoint.clone();
-        self.pool.execute(move || {
-            let timer = SlowTimer::new();
-            let tp = req.get_tp();
-            end_point.handle_request(req, on_resp);
-            slow_log!(timer, "request type {} takes {:?}", tp, timer.elapsed());
-        });
+pub struct RequestTask(Request, OnResponse);
+
+impl RequestTask {
+    pub fn new(req: Request, on_resp: OnResponse) -> RequestTask {
+        RequestTask(req, on_resp)
+    }
+}
+
+impl Display for RequestTask {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f,
+               "request [context {:?}, tp: {}, ranges: {:?}]",
+               self.0.get_context(),
+               self.0.get_tp(),
+               self.0.get_ranges())
+    }
+}
+
+impl BatchRunnable<RequestTask> for EndPointHost {
+    #[allow(for_kv_map)]
+    fn run_batch(&mut self, reqs: &mut Vec<RequestTask>) {
+        let mut grouped_reqs = map![];
+        for req in reqs.drain(..) {
+            let key = {
+                let ctx = req.0.get_context();
+                (ctx.get_region_id(),
+                 ctx.get_region_epoch().get_conf_ver(),
+                 ctx.get_region_epoch().get_version(),
+                 ctx.get_peer().get_id(),
+                 ctx.get_peer().get_store_id())
```

Comment:
What if `store_id` is wrong? Only `region_id` is not enough, may contains stale epoch.